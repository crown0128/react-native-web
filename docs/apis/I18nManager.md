# I18nManager

Control and set the layout and writing direction of the application.

## Properties

**isRTL**: bool = false

Whether the application is currently in RTL mode.

## Methods

static **allowRTL**(allowRTL: bool)

Allow the application to display in RTL mode.

static **forceRTL**(forceRTL: bool)

Force the application to display in RTL mode.

static **setPreferredLanguageRTL**(isRTL: bool)

Set the application's preferred writing direction to RTL. You will need to
determine the user's preferred locale server-side (from HTTP headers) and
decide whether it's an RTL language.
