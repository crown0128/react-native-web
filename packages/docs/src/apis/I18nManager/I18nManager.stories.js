export default {
  title: 'APIs|I18nManager',
  includeStories: []
};

export { default as layoutRTL } from './examples/LayoutRTL';
