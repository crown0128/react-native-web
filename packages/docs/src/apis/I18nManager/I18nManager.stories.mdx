import { Meta, Props, Preview, Story } from '@storybook/addon-docs/blocks';
import * as Stories from './I18nManager.stories.js';

<Meta title="APIs|I18nManager" />

# I18nManager

Control and query the layout and writing direction of the application.

## Properties

### isRTL

Whether the application is currently in RTL mode.

### doLeftAndRightSwapInRTL

Whether the application swaps left/right styles in RTL mode. (Default is `true`
but will eventually be changed to `false`).

## Methods

### allowRTL(bool)

Allow the application to display in RTL mode.

### forceRTL(bool)

Force the application to display in RTL mode.

### swapLeftAndRightInRTL(bool)

Control whether the application swaps `left`/`right` styles in RTL mode. It is
recommended that applications rely on `start`/`end` styles and disable automatic
BiDi-flipping of `left`/`right` styles, as this will eventually become the
default.

### setPreferredLanguageRTL(bool)

Set the application's preferred writing direction to RTL. You may need to infer
the user's preferred locale on the server (from HTTP headers) and decide whether
it's an RTL language. (Web-only)

## Example

<Preview withSource='none'>
  <Story inline={false} name="layoutRTL">
    <Stories.layoutRTL />
  </Story>
</Preview>
