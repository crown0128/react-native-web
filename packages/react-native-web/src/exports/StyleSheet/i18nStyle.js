/**
 * Copyright (c) 2016-present, Nicolas Gallagher.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import I18nManager from '../I18nManager';
import multiplyStyleLengthValue from '../../modules/multiplyStyleLengthValue';

const emptyObject = {};

const borderTopLeftRadius = 'borderTopLeftRadius';
const borderTopRightRadius = 'borderTopRightRadius';
const borderBottomLeftRadius = 'borderBottomLeftRadius';
const borderBottomRightRadius = 'borderBottomRightRadius';
const borderLeftColor = 'borderLeftColor';
const borderLeftStyle = 'borderLeftStyle';
const borderLeftWidth = 'borderLeftWidth';
const borderRightColor = 'borderRightColor';
const borderRightStyle = 'borderRightStyle';
const borderRightWidth = 'borderRightWidth';
const right = 'right';
const marginLeft = 'marginLeft';
const marginRight = 'marginRight';
const paddingLeft = 'paddingLeft';
const paddingRight = 'paddingRight';
const left = 'left';

// Map of LTR property names to their BiDi equivalent.
const PROPERTIES_FLIP = {
  borderTopLeftRadius: borderTopRightRadius,
  borderTopRightRadius: borderTopLeftRadius,
  borderBottomLeftRadius: borderBottomRightRadius,
  borderBottomRightRadius: borderBottomLeftRadius,
  borderLeftColor: borderRightColor,
  borderLeftStyle: borderRightStyle,
  borderLeftWidth: borderRightWidth,
  borderRightColor: borderLeftColor,
  borderRightStyle: borderLeftStyle,
  borderRightWidth: borderLeftWidth,
  left: right,
  marginLeft: marginRight,
  marginRight: marginLeft,
  paddingLeft: paddingRight,
  paddingRight: paddingLeft,
  right: left
};

// Map of I18N property names to their LTR equivalent.
const PROPERTIES_I18N = {
  borderTopStartRadius: borderTopLeftRadius,
  borderTopEndRadius: borderTopRightRadius,
  borderBottomStartRadius: borderBottomLeftRadius,
  borderBottomEndRadius: borderBottomRightRadius,
  borderStartColor: borderLeftColor,
  borderStartStyle: borderLeftStyle,
  borderStartWidth: borderLeftWidth,
  borderEndColor: borderRightColor,
  borderEndStyle: borderRightStyle,
  borderEndWidth: borderRightWidth,
  end: right,
  marginStart: marginLeft,
  marginEnd: marginRight,
  paddingStart: paddingLeft,
  paddingEnd: paddingRight,
  start: left
};

const PROPERTIES_VALUE = {
  clear: true,
  float: true,
  textAlign: true
};

// Invert the sign of a numeric-like value
const additiveInverse = (value: String | Number) => multiplyStyleLengthValue(value, -1);

// Convert I18N properties and values
const convertProperty = (prop: String): String => {
  return PROPERTIES_I18N.hasOwnProperty(prop) ? PROPERTIES_I18N[prop] : prop;
};
const convertValue = (value: String): String => {
  return value === 'start' ? 'left' : value === 'end' ? 'right' : value;
};

// BiDi flip properties and values
const flipProperty = (prop: String): String => {
  return PROPERTIES_FLIP.hasOwnProperty(prop) ? PROPERTIES_FLIP[prop] : prop;
};
const flipValue = (value: String): String => {
  return value === 'left' ? 'right' : value === 'right' ? 'left' : value;
};

const i18nStyle = originalStyle => {
  const isRTL = I18nManager.isRTL;

  const style = originalStyle || emptyObject;
  const nextStyle = {};
  const frozenProps = {};

  for (const originalProp in style) {
    if (!Object.prototype.hasOwnProperty.call(style, originalProp)) {
      continue;
    }

    let prop = originalProp;
    let value = style[originalProp];
    let shouldFreezeProp = false;

    // Process I18N properties and values
    if (PROPERTIES_I18N[prop]) {
      prop = convertProperty(prop);
      // I18N properties takes precendence over left/right
      shouldFreezeProp = true;
    } else if (PROPERTIES_VALUE[prop]) {
      value = convertValue(value);
    }

    if (isRTL) {
      if (PROPERTIES_FLIP[prop]) {
        const newProp = flipProperty(prop);
        if (!frozenProps[prop]) {
          nextStyle[newProp] = value;
        }
      } else if (PROPERTIES_VALUE[prop]) {
        nextStyle[prop] = flipValue(value);
      } else if (prop === 'textShadowOffset') {
        nextStyle[prop] = value;
        nextStyle[prop].width = additiveInverse(value.width);
      } else {
        nextStyle[prop] = style[prop];
      }
    } else {
      if (!frozenProps[prop]) {
        nextStyle[prop] = value;
      }
    }

    // Mark the style prop as frozen
    if (shouldFreezeProp) {
      frozenProps[prop] = true;
    }
  }

  return nextStyle;
};

export default i18nStyle;
