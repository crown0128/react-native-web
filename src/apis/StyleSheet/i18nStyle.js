import I18nManager from '../I18nManager';
import multiplyStyleLengthValue from '../../modules/multiplyStyleLengthValue';

const emptyObject = {};

/**
 * Map of property names to their BiDi equivalent.
 */
const PROPERTIES_TO_SWAP = {
  borderTopLeftRadius: 'borderTopRightRadius',
  borderTopRightRadius: 'borderTopLeftRadius',
  borderBottomLeftRadius: 'borderBottomRightRadius',
  borderBottomRightRadius: 'borderBottomLeftRadius',
  borderLeftColor: 'borderRightColor',
  borderLeftStyle: 'borderRightStyle',
  borderLeftWidth: 'borderRightWidth',
  borderRightColor: 'borderLeftColor',
  borderRightWidth: 'borderLeftWidth',
  borderRightStyle: 'borderLeftStyle',
  left: 'right',
  marginLeft: 'marginRight',
  marginRight: 'marginLeft',
  paddingLeft: 'paddingRight',
  paddingRight: 'paddingLeft',
  right: 'left'
};

const PROPERTIES_SWAP_LEFT_RIGHT = {
  clear: true,
  float: true,
  textAlign: true
};

/**
 * Invert the sign of a numeric-like value
 */
const additiveInverse = (value: String | Number) => multiplyStyleLengthValue(value, -1);

/**
 * BiDi flip the given property.
 */
const flipProperty = (prop: String): String => {
  return PROPERTIES_TO_SWAP.hasOwnProperty(prop) ? PROPERTIES_TO_SWAP[prop] : prop;
};

/**
 * BiDi flip translateX
 */
const flipTransform = (transform: Object): Object => {
  const translateX = transform.translateX;
  if (translateX != null) {
    transform.translateX = additiveInverse(translateX);
  }
  return transform;
};

const swapLeftRight = (value: String): String => {
  return value === 'left' ? 'right' : value === 'right' ? 'left' : value;
};

const i18nStyle = originalStyle => {
  if (!I18nManager.isRTL) {
    return originalStyle;
  }

  const style = originalStyle || emptyObject;
  const nextStyle = {};

  for (const prop in style) {
    if (!Object.prototype.hasOwnProperty.call(style, prop)) {
      continue;
    }

    const value = style[prop];

    if (PROPERTIES_TO_SWAP[prop]) {
      const newProp = flipProperty(prop);
      nextStyle[newProp] = value;
    } else if (PROPERTIES_SWAP_LEFT_RIGHT[prop]) {
      nextStyle[prop] = swapLeftRight(value);
    } else if (prop === 'textShadowOffset') {
      nextStyle[prop] = value;
      nextStyle[prop].width = additiveInverse(value.width);
    } else if (prop === 'transform' && Array.isArray(value)) {
      nextStyle[prop] = style[prop].map(flipTransform);
    } else {
      nextStyle[prop] = style[prop];
    }
  }

  return nextStyle;
};

module.exports = i18nStyle;
