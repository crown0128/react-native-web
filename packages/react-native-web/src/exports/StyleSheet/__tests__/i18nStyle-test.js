/* eslint-env jasmine, jest */

import I18nManager from '../../I18nManager';
import i18nStyle from '../i18nStyle';

const styleLeftRight = {
  borderLeftColor: 'red',
  borderRightColor: 'blue',
  borderTopLeftRadius: 10,
  borderTopRightRadius: '1rem',
  borderBottomLeftRadius: 20,
  borderBottomRightRadius: '2rem',
  borderLeftStyle: 'solid',
  borderRightStyle: 'dotted',
  borderLeftWidth: 5,
  borderRightWidth: 6,
  left: 1,
  marginLeft: 7,
  marginRight: 8,
  paddingLeft: 9,
  paddingRight: 10,
  right: 2,
  textAlign: 'left',
  textShadowOffset: { width: '1rem', height: 10 }
};

const styleStartEnd = {
  borderStartColor: 'red',
  borderEndColor: 'blue',
  borderTopStartRadius: 10,
  borderTopEndRadius: '1rem',
  borderBottomStartRadius: 20,
  borderBottomEndRadius: '2rem',
  borderStartStyle: 'solid',
  borderEndStyle: 'dotted',
  borderStartWidth: 5,
  borderEndWidth: 6,
  start: 1,
  marginStart: 7,
  marginEnd: 8,
  paddingStart: 9,
  paddingEnd: 10,
  end: 2,
  textAlign: 'start',
  textShadowOffset: { width: '1rem', height: 10 }
};

describe('StyleSheet/i18nStyle', () => {
  describe('LTR mode', () => {
    beforeEach(() => {
      I18nManager.allowRTL(false);
    });

    afterEach(() => {
      I18nManager.allowRTL(true);
    });

    test("doesn't flip left/right", () => {
      expect(i18nStyle(styleLeftRight)).toMatchSnapshot();
    });

    test("converts and doesn't flip start/end", () => {
      expect(i18nStyle(styleStartEnd)).toMatchSnapshot();
    });

    test('start/end takes precedence over left/right', () => {
      const style = {
        borderTopStartRadius: 10,
        borderTopLeftRadius: 0,
        end: 10,
        right: 0,
        marginStart: 10,
        marginLeft: 0
      };
      const expected = {
        borderTopLeftRadius: 10,
        marginLeft: 10,
        right: 10
      };
      expect(i18nStyle(style)).toEqual(expected);
    });
  });

  describe('RTL mode', () => {
    beforeEach(() => {
      I18nManager.forceRTL(true);
    });

    afterEach(() => {
      I18nManager.forceRTL(false);
    });

    test('flips left/right', () => {
      expect(i18nStyle(styleLeftRight)).toMatchSnapshot();
    });

    test('converts and flips start/end', () => {
      expect(i18nStyle(styleStartEnd)).toMatchSnapshot();
    });

    test('start/end takes precedence over left/right', () => {
      const style = {
        borderTopStartRadius: 10,
        borderTopLeftRadius: 0,
        end: 10,
        right: 0,
        marginStart: 10,
        marginLeft: 0
      };
      const expected = {
        borderTopRightRadius: 10,
        marginRight: 10,
        left: 10
      };
      expect(i18nStyle(style)).toEqual(expected);
    });
  });
});
