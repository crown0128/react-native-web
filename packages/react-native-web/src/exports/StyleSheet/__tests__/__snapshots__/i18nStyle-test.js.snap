// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheet/i18nStyle LTR mode converts and doesn't flip start/end 1`] = `
Object {
  "borderBottomLeftRadius": 20,
  "borderBottomRightRadius": "2rem",
  "borderLeftColor": "red",
  "borderLeftStyle": "solid",
  "borderLeftWidth": 5,
  "borderRightColor": "blue",
  "borderRightStyle": "dotted",
  "borderRightWidth": 6,
  "borderTopLeftRadius": 10,
  "borderTopRightRadius": "1rem",
  "left": 1,
  "marginLeft": 7,
  "marginRight": 8,
  "paddingLeft": 9,
  "paddingRight": 10,
  "right": 2,
  "textAlign": "left",
  "textShadowOffset": Object {
    "height": 10,
    "width": "1rem",
  },
}
`;

exports[`StyleSheet/i18nStyle LTR mode doesn't flip left/right 1`] = `
Object {
  "borderBottomLeftRadius": 20,
  "borderBottomRightRadius": "2rem",
  "borderLeftColor": "red",
  "borderLeftStyle": "solid",
  "borderLeftWidth": 5,
  "borderRightColor": "blue",
  "borderRightStyle": "dotted",
  "borderRightWidth": 6,
  "borderTopLeftRadius": 10,
  "borderTopRightRadius": "1rem",
  "left": 1,
  "marginLeft": 7,
  "marginRight": 8,
  "paddingLeft": 9,
  "paddingRight": 10,
  "right": 2,
  "textAlign": "left",
  "textShadowOffset": Object {
    "height": 10,
    "width": "1rem",
  },
}
`;

exports[`StyleSheet/i18nStyle RTL mode converts and flips start/end 1`] = `
Object {
  "borderBottomLeftRadius": "2rem",
  "borderBottomRightRadius": 20,
  "borderLeftColor": "blue",
  "borderLeftStyle": "dotted",
  "borderLeftWidth": 6,
  "borderRightColor": "red",
  "borderRightStyle": "solid",
  "borderRightWidth": 5,
  "borderTopLeftRadius": "1rem",
  "borderTopRightRadius": 10,
  "left": 2,
  "marginLeft": 8,
  "marginRight": 7,
  "paddingLeft": 10,
  "paddingRight": 9,
  "right": 1,
  "textAlign": "right",
  "textShadowOffset": Object {
    "height": 10,
    "width": "-1rem",
  },
}
`;

exports[`StyleSheet/i18nStyle RTL mode flips left/right 1`] = `
Object {
  "borderBottomLeftRadius": "2rem",
  "borderBottomRightRadius": 20,
  "borderLeftColor": "blue",
  "borderLeftStyle": "dotted",
  "borderLeftWidth": 6,
  "borderRightColor": "red",
  "borderRightStyle": "solid",
  "borderRightWidth": 5,
  "borderTopLeftRadius": "1rem",
  "borderTopRightRadius": 10,
  "left": 2,
  "marginLeft": 8,
  "marginRight": 7,
  "paddingLeft": 10,
  "paddingRight": 9,
  "right": 1,
  "textAlign": "right",
  "textShadowOffset": Object {
    "height": 10,
    "width": "-1rem",
  },
}
`;
